client/src/main/java/io/pravega/client/stream/impl/TxnKeepAlive.java

```diff
@@ -0,0 +1,114 @@
+
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import io.pravega.client.stream.EventWriterConfig;
+import io.pravega.client.stream.Stream;
+import io.pravega.common.concurrent.Futures;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.Synchronized;
+import lombok.extern.slf4j.Slf4j;
+
+import static io.pravega.common.concurrent.Futures.getException;
+
+/**
+ * TxnKeepAlive is used to send KeepAlives to renew the transaction lease for active transactions.
+ * It invokes io.pravega.client.stream.impl.Controller#pingTransaction() on the controller to renew the lease of active
+ * transactions. Incase of a controller instance not being reachable the controller client takes care of retrying on a
+ * different controller instance see io.pravega.client.stream.impl.ControllerResolverFactory for details.
+ */
+@Slf4j
+public class TxnKeepAlive implements AutoCloseable {
+    private static final double KEEP_ALIVE_INTERVAL_FACTOR = 0.7; //keepalive interval = factor * txn lease time.
+
+    private final Stream stream;
+    private final Controller controller;
+    private final long txnLeaseMillis;
+    private final long pingIntervalMillis;
+    private final ScheduledExecutorService executor;
+    private final List<UUID> txnList = new CopyOnWriteArrayList<>();
+    private final AtomicBoolean isStarted = new AtomicBoolean();
+    @GuardedBy("$lock")
+    private ScheduledFuture<?> pingTxnTask;
+
+    TxnKeepAlive(EventWriterConfig config, Stream stream, Controller controller, ScheduledExecutorService
+            internalExecutor) {
+        this.txnLeaseMillis = config.getTransactionTimeoutTime();
+        this.pingIntervalMillis = getKeepAliveInterval(txnLeaseMillis);
+        this.stream = stream;
+        this.controller = controller;
+        this.executor = internalExecutor;
+    }
+
+    void startTxnKeepAlive(UUID txnID) {
+        txnList.add(txnID);
+        startPeriodicPingTxn();
+    }
+
+    void stopTxnKeepAlive(UUID txnID) {
+        txnList.remove(txnID);
+    }
+
+    private long getKeepAliveInterval(long txnLeaseMillis) {
+        double keepAliveInterval = txnLeaseMillis * KEEP_ALIVE_INTERVAL_FACTOR;
+        if (keepAliveInterval < TimeUnit.SECONDS.toMillis(10)) {
+            log.warn("Transaction Keep Alive interval is less than 10 seconds(lower bound)");
+        }
+        //Ping interval cannot be less than KeepAlive task interval of 10seconds.
+        return Math.max(TimeUnit.SECONDS.toMillis(10), (long) keepAliveInterval);
+    }
+
+    @Synchronized
+    private void startPeriodicPingTxn() {
+        if (!isStarted.get()) {
```

Comment:
You don't need `isStarted`. You can check if `pingTxnTask` is null or not.

Reply:
yes, updated the code.